abi/abi_encode.go

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2018 De-labtory
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package abi
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+	"strconv"
+
+	"github.com/DE-labtory/koa/crpyto"
+	"github.com/ethereum/go-ethereum/common/math"
+)
+
+type Pointer []byte
+type Size []byte
+type Value []byte
+
+// Encode abi parameters
+func Encode(params ...interface{}) ([]byte, error) {
+	pointers := make([]Pointer, len(params))
+	sizes := make([]Size, len(params))
+	values := make([]Value, len(params))
+
+	for index, param := range params {
+		bytesValue, err := EncodeOperand(param)
+		if err != nil {
+			return nil, err
+		}
+		values[index] = append(values[index], bytesValue...)
+
+		size, err := EncodeOperand(len(bytesValue))
+		if err != nil {
+			return nil, err
+		}
+
+		sizes[index] = left4padding(size)
+	}
+
+	for index := 0; index < len(params); index++ {
+		length := len(params) * 4
+
+		for from := 0; from < index; from++ {
+			length += len(sizes[from]) + len(values[from])
+		}
+
+		pointer, err := EncodeOperand(length)
+		if err != nil {
+			return nil, err
+		}
+
+		pointers[index] = left4padding(pointer)
+	}
+
+	var Args []byte
+
+	for _, pointer := range pointers {
+		Args = append(Args, pointer...)
+	}
+
+	for index := 0; index < len(params); index++ {
+		Args = append(Args, sizes[index]...)
+		Args = append(Args, values[index]...)
+	}
+
+	return Args, nil
```

Comment:
Encode is so big. Can you divide it?

Reply:
It would be better to divide them by `for` statement.

Ex)
```go
func Encode(params ...interface{}) ([]byte, error) {
  pointers := make([]Pointer, len(params))
  sizes := make([]Size, len(params))
  values := make([]Value, len(params)) 

  encodeValue(sizes, values)

  encodePointer(pointers)

  var Args []byte

  for _, pointer := range pointers {
    Args = append(Args, pointer...)
  }

  for index := 0; index < len(params); index++ {
    Args = append(Args, sizes[index]...)
    Args = append(Args, values[index]...)
  }
  
  return Args, nil
}